### PR TITLE
Graph decoration performance and other improvements (4.0)

### PR DESCRIPTION
**Pull request by jharwig**
- [x] joeferner
- [x] mwizeman joeybrk372 jharwig sfeng88

* Decoraction performance issues
* Extract previewMD5 from within the product store and put in separate
area to preview loops of react render updates
* Fix decoration data methods with promises
* Ancillary nodes don't participate in layouts
* Fix focus dimming of decorations
* Removes viewport updating that would try to return the pan/zoom when
switching products, was not correctly and caused excessive renders.
Simply return to fitting the graph until we can actually persist
viewports across sessions
* Fix layouts with decorations acting erratictly
* Fix diff panel js error
* Replace circle layout with concentric circles based on node degree (number of edges)
* Add value types to deepObjectCache tree

Points of Regression: Removes viewport saving when switching graph products

CHANGELOG
Changed: Graph circle layout has been replaced with concentric circles layout with nodes having most edges in the center
Fixed: Better performance with graphs that have node decorations
Fixed: Graph decorations with data functions that return promises
Fixed: Graph layouts when nodes have decorations would perform erratically

### CODE REVIEW

**Comment by jharwig** [comment]
Will make master PR after reviewed

File: web/plugins/graph-product/src/main/resources/org/visallo/web/product/graph/Cytoscape.jsx [hunk 1]
```diff
@@ -488,24 +452,30 @@ define([
...
+            // parents (because we don't want those to layout, so just hack
+            // them to not show parents.
+            elements.forEach(({_private: el }) => {
+                el._restoreParent = el.parent;
+                el._restoreDataParent = el.data.parent;
+                el.parent = null;
+                el.data.parent = null;
+            })
+            elements.layout(opts).run();
+            elements.forEach(({ _private: el }) => {
```
[the remark this answers is not in the document]
Comment (jharwig) [reply web/plugins/graph-product/src/main/resources/org/visallo/web/product/graph/Cytoscape.jsx]: Yup, but only the constructor needs the fix. I wanted to revert as quickly as possible. Plus decorations won't animate with the nodes if I wait.

File: web/plugins/graph-product/src/main/resources/org/visallo/web/product/graph/Cytoscape.jsx [hunk 2]
```diff
@@ -488,24 +452,30 @@ define([
...
-            if (onlySelected) {
-                elements.layout(opts).run();
-            } else {
-                cy.layout(opts).run();
-            }
-
+            // Force directed has issues if we don't pass the decoration
+            // parents (because we don't want those to layout, so just hack
+            // them to not show parents.
+            elements.forEach(({_private: el }) => {
```
[the remark this answers is not in the document]
Comment (jharwig) [reply web/plugins/graph-product/src/main/resources/org/visallo/web/product/graph/Cytoscape.jsx]: done

**Comment by sfeng88** [comment]
@jharwig please make a master PR. merging this one now.

**Comment by jharwig** [comment]
done

**Pull request merged**